Fix duplicate rules

notnyt opened the pull request:
Every time read_rules is called, it was creating a duplicate entry in the categories array

notnyt commented:
There's still something wrong here, and an issue with local rules not getting added.  Working on another patch to get it all sorted, hopefully soon.

shirkdog commented:
I have identified an issue with ignoring rules, but not the local rules. Did this just crop up for you? as there has been a number of pull requests that I have brought in and I have only checked for sanity. We need all bad things captured in new issues.

notnyt commented:
local rules weren't having their sid written into the $categories hash, so when rule_category_write was called, they'd come up empty with just the gid header.  I have it working now, however it's adding a GID:1 based rules header and matching sid for some reason from an ET file.  Will get this sorted and get the fix pushed.

````
(DEBIAN)root@rooter:/etc/suricata/pp# cat > ../rules/iprep.rules
drop ip any any -> any any (msg:"IPREP Blacklist Match"; iprep:src,CnC,=,99; sid:1; rev:1;)

(DEBIAN)root@rooter:/etc/suricata/pp# /etc/suricata/pp/pulledpork.pl -c /etc/suricata/pp/pulledpork.conf -T -S suricata -k

[...]


(DEBIAN)root@rooter:/etc/suricata/pp# cat ../rules/iprep.rules


# ----- Begin iprep Rules Category ----- #

# -- Begin GID:0 Based Rules -- #

drop ip any any -> any any (msg:"IPREP Blacklist Match"; iprep:src,CnC,=,99; sid:1; rev:1;)

# -- Begin GID:1 Based Rules -- #

#alert http any any -> any any (msg:"FILEEXT JPG file claimed"; fileext:"jpg"; sid:1; rev:1;)

````

notnyt commented:
Looks like that wasn't a bug, just cruft from a file from troubleshooting earlier.  However it looks like there's a multiline processing bug for local files, so will get that fixed instead ;)

notnyt commented:
Some of the further trouble I'm seeing is read_rules being called on $rule_file_path in order to do sid changelog.  This causes entries to be created in the categories hash, which causes empty files to be written during rule_category_writes.  Going to just add an option to prevent the categories hash from being touched in this case.

notnyt commented:
There's also a bug where the local files get read a second time if they're in the rules dir and it processes them as GID 1.  Adding a sanity check in read_rules to not read local files when reading all entries in $path

notnyt commented:
pushing a patch in a minute, just running some more test.

- Converted the array of sids to a hash since its much more efficient and no dupes
- Added a check to not add old files to the categories hash, this prevents them from being blanked
- Added a check to not parse local files when iterating over a rules directory since they're handled separately
